### PR TITLE
Do not trigger 2PC for reads on local execution

### LINKED ISSUE #4777: Unexpected error on standby nodes when distributed subplans exists
Queries with intermediate results fail:
```SQL
WITH cte_1 AS (SELECT * FROM test LIMIT 5) SELECT * FROM cte_1;
ERROR:  cannot assign TransactionIds during recovery

 SELECT * FROM (SELECT * FROM test LIMIT 5) as foo;
ERROR:  cannot assign TransactionIds during recovery
Time: 9.145 ms

```

Though regular queries work fine:
```SQL
SELECT * FROM test LIMIT 5;
 a  
----
  1
 15
 25
 63
 82
(5 rows)

Time: 42.151 ms
```

### PR DESCRIPTION
**Pull request by onderkalaci**
Fixes #4777 

Before this commit, Citus used 2PC no matter what kind of
local query execution happens.

For example, if the coordinator has shards (and the workers as well),
even a simple SELECT query could start 2PC:
```SQL

WITH cte_1 AS (SELECT * FROM test LIMIT 10) SELECT count(*) FROM cte_1;
```

In this query, the local execution of the shards (and also intermediate
result reads) triggers the 2PC.

To prevent that, Citus now distinguishes local reads and local writes.
And, Citus switches to 2PC only if a modification happens. This may
still lead to unnecessary 2PCs when there is a local modification
and remote SELECTs only. Though, we are going to handle that separately
via #4587.

### CODE REVIEW

File: src/backend/distributed/executor/adaptive_executor.c [hunk 1]
```diff
@@ -4341,7 +4341,11 @@ PlacementExecutionDone(TaskPlacementExecution *placementExecution, bool succeede
 		 * As we decided to failover this task to local execution, we cannot
 		 * allow remote execution to this pool during this distributedExecution.
 		 */
-		SetLocalExecutionStatus(LOCAL_EXECUTION_REQUIRED);
+		LocalExecutionStatus status =
+			ReadOnlyTask(task->taskType) ? LOCAL_EXECUTION_REQUIRED_READONLY :
```
Comment (onderkalaci) [inline src/backend/distributed/executor/adaptive_executor.c:4345]: I'll try to add a regression test for this part as well

File: src/backend/distributed/executor/local_executor.c [hunk 2]
```diff
@@ -839,7 +843,7 @@ TaskAccessesLocalNode(Task *task)
 void
 ErrorIfTransactionAccessedPlacementsLocally(void)
 {
-	if (GetCurrentLocalExecutionStatus() == LOCAL_EXECUTION_REQUIRED)
+	if (GetCurrentLocalExecutionStatus() >= LOCAL_EXECUTION_REQUIRED_READONLY)
```
Comment (marcocitus) [inline src/backend/distributed/executor/local_executor.c:846]: The naming of LOCAL_EXECUTION_REQUIRED_ seems a bit strange. It's not very intuitive what REQUIRED_MODIFY means and that one status is greater than the other. 

Could this be a separate flag, or are there some better names?

Comment (onderkalaci) [reply src/backend/distributed/executor/local_executor.c]: Agreed both (a) the naming was not great (b) we should not have an ordering between the status.

I initially tried having one more state, but that seemed more complicated. Now, I pushed a WIP commit that improves (a) and (b). 

Does that look or do you think we should introduce a new state?

File: src/backend/distributed/commands/alter_table.c [hunk 3]
```diff
@@ -1700,7 +1700,7 @@ SwitchToSequentialAndLocalExecutionIfShardNameTooLong(char *relationName,
 				 longestShardName, relationName);
 
 			SetLocalMultiShardModifyModeToSequential();
-			SetLocalExecutionStatus(LOCAL_EXECUTION_REQUIRED);
+			SetLocalExecutionStatus(LOCAL_EXECUTION_MODIFY_ACCESSED);
```
Comment (marcocitus) [inline src/backend/distributed/commands/alter_table.c:1703]: Maybe something like LOCAL_EXECUTION_PERFORMED_MODIFICATION?

File: src/test/regress/expected/coordinator_shouldhaveshards.out [hunk 4]
```diff
@@ -895,6 +895,145 @@ NOTICE:  executing the command locally: SELECT count(*) AS count FROM (SELECT in
...
+NOTICE:  executing the command locally: SELECT count(*) AS count FROM coordinator_shouldhaveshards.test_1503003 test WHERE true
+ count
+---------------------------------------------------------------------
+     1
+(1 row)
+
+  SELECT coordinated_transaction_uses_2PC();
+ coordinated_transaction_uses_2pc
+---------------------------------------------------------------------
+ t
```
Comment (marcocitus) [inline src/test/regress/expected/coordinator_shouldhaveshards.out:1018]: should we merge https://github.com/citusdata/citus/pull/4587 and rebase?

Comment (onderkalaci) [reply src/test/regress/expected/coordinator_shouldhaveshards.out:1018]: I think this test wouldn't change, but we would prevent actually  doing 2PC for remote connections.

I'd rather merge this and test that on top of this as well, if that works?

Comment (marcocitus) [reply src/test/regress/expected/coordinator_shouldhaveshards.out:1018]: Why do we have 2PC here then?

Comment (marcocitus) [reply src/test/regress/expected/coordinator_shouldhaveshards.out:1018]: I wonder if we should rename the variable to ShouldUse2PC

File: src/backend/distributed/executor/adaptive_executor.c [hunk 5]
```diff
@@ -1165,16 +1166,16 @@ DecideTransactionPropertiesForTaskList(RowModifyLevel modLevel, List *taskList,
...
 		/*
-		 * In case localExecutionHappened, we force the executor to use 2PC.
-		 * The primary motivation is that at this point we're definitely expanding
-		 * the nodes participated in the transaction. And, by re-generating the
-		 * remote task lists during local query execution, we might prevent the adaptive
-		 * executor to kick-in 2PC (or even start coordinated transaction, that's why
-		 * we prefer adding this check here instead of
-		 * Activate2PCIfModifyingTransactionExpandsToNewNode()).
+		 * In case a local execution with modification happened, we force the
+		 * executor to use 2PC. The primary motivation is that at this point
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c:1173]: do we also need to do this if this command is not a modification?

Comment (marcocitus) [reply src/backend/distributed/executor/adaptive_executor.c:1173]: (if not, we should probably change the name to *Should*Use2PC)

File: src/test/regress/expected/coordinator_shouldhaveshards.out (hunk 4 above)
Comment (onderkalaci) [reply src/test/regress/expected/coordinator_shouldhaveshards.out:1018]: Or, if I follow your other comment, this should be false. I think that makes sense

File: src/backend/distributed/executor/adaptive_executor.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:1174]: With SELECT FOR UPDATE, we do not use 2PC. As far as I can say, that is expected behaviour, right?

Comment (marcocitus) [reply src/backend/distributed/executor/adaptive_executor.c:1174]: Seems fine to me..

**Comment by marcocitus** [comment]
We seem to not be doing 2PC in this scenario:
```
postgres=# begin;
BEGIN
Time: 1.834 ms
postgres=*# insert into test values (1,2);
NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 2, '2021-03-09 09:31:18.133049+01');
DETAIL:  on server marco@localhost:1302 connectionId: 3
NOTICE:  issuing INSERT INTO public.test_102372 (x, y) VALUES (1, 2)
DETAIL:  on server marco@localhost:1302 connectionId: 3
INSERT 0 1
Time: 4.217 ms
postgres=*# insert into test values (2,2);
NOTICE:  executing the command locally: INSERT INTO public.test_102400 (x, y) VALUES (2, 2)
INSERT 0 1
Time: 2.074 ms
postgres=*# end;
NOTICE:  issuing COMMIT
DETAIL:  on server marco@localhost:1302 connectionId: 3
COMMIT
```

**Comment by onderkalaci** [comment]
The other way is also broken, but at least not dangerous:

```SQL
BEGIN;
Time: 0.171 ms
insert into test VALUES (2,2);
NOTICE:  executing the command locally: INSERT INTO public.test_102032 (a, b) VALUES (2, 2)
INSERT 0 1
Time: 0.588 ms
insert into test VALUES (2,2);
NOTICE:  executing the command locally: INSERT INTO public.test_102032 (a, b) VALUES (2, 2)
INSERT 0 1
Time: 0.569 ms

select count(*) from test;
NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 6, '2021-03-09 20:15:36.54673+01');
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102008 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102009 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102010 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102011 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102012 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102013 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102014 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102015 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102016 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102017 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102018 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102019 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102020 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102021 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102022 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT count(*) AS count FROM test_102023 test WHERE true
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102024 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102025 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102026 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102027 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102028 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102029 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102030 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102031 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102032 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102033 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102034 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102035 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102036 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102037 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102038 test WHERE true
NOTICE:  executing the command locally: SELECT count(*) AS count FROM test_102039 test WHERE true
 count 
-------
 33826
(1 row)

COMMIT;
NOTICE:  issuing PREPARE TRANSACTION 'citus_0_61899_6_2'
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing COMMIT PREPARED 'citus_0_61899_6_2'
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
COMMIT
Time: 2.805 ms
```

**Comment by onderkalaci** [comment]
> We seem to not be doing 2PC in this scenario:

@marcocitus 

My new proposal is here  (on top of #4587): https://github.com/citusdata/citus/compare/fix_2pc_local_exec?expand=1

We basically always enable 2PC after any local modifications. If remote connections do not  have any modifications, we already skip them via #4587.  So, in a sense, we treat enabling 2PC as a related but separate concept than doing 2PC over remote connections. In fact, with #4587 we could probably just treat every coordinated transaction as 2PC, and let the logic in #4587 to decide what to do. Though, lets not do that today :)

**Comment by onderkalaci** [comment]
superseded by #4806

**Pull request closed without merging**